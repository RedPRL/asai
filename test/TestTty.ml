open Asai


let width = 10
let template = "abcdefghijklmnopqrstuvw"

let s title : Range.source =
  `String
    { title = title
    ; content =
        String.concat "\n" @@
        List.init  (String.length template) @@ fun i ->
        String.make width template.[i]
    }

let s1 = s None
let s2 = s (Some "/path/to/file.cool")

let (~@) source line col : Range.position =
  { source
  ; offset = (line-1) * (width+1) + col
  ; start_of_line = (line-1) * (width+1)
  ; line_num = line}

module Reporter =
struct
  module Code =
  struct
    type t = Hello | Bye
    let default_severity : t -> Asai.Diagnostic.severity =
      function
      | Hello -> Warning
      | Bye -> Error
    let short_code : t -> string =
      function
      | Hello -> "hello"
      | Bye -> "bye"
  end

  include Asai.Reporter.Make(Code)
end

module Terminal = Asai.Tty.Make(Reporter.Code)

let () =
  let handler =
    match Sys.argv with
    | [| _ ; "--interact" |] -> fun d -> Terminal.interact d
    | _ -> fun d -> Terminal.display ~terminal_capacity:Notty.Cap.ansi d
  in
  Reporter.run ~emit:handler ~fatal:handler @@ fun () ->
  Reporter.emitf Hello "aloha %d" 100;
  Reporter.emitf Bye "aloha %d" 200;
  Reporter.emit ~loc:(Range.make (~@ s1 1 3, ~@ s1 2 4)) Hello "hello here!";
  Reporter.emit ~loc:(Range.make (~@ s2 2 3, ~@ s2 3 5)) Bye "bye there!";
  Reporter.emit ~loc:(Range.make (~@ s1 1 3, ~@ s1 2 4)) Hello "this is a bug" ~severity:Bug;
  Reporter.emit ~loc:(Range.make (~@ s1 1 3, ~@ s1 2 4)) Hello "this is an error" ~severity:Error;
  Reporter.emit ~loc:(Range.make (~@ s1 1 3, ~@ s1 2 4)) Hello "this is a warning" ~severity:Warning;
  Reporter.emit ~loc:(Range.make (~@ s1 1 3, ~@ s1 2 4)) Hello "this is an info" ~severity:Info;
  Reporter.emit ~loc:(Range.make (~@ s1 1 3, ~@ s1 2 4)) Hello "this is a hint" ~severity:Hint;
  Reporter.emit ~loc:(Range.make (~@ s2 1 3, ~@ s2 3 4)) Hello "this is a bug" ~severity:Bug;
  Reporter.emit ~loc:(Range.make (~@ s2 1 3, ~@ s2 3 4)) Hello "this is an error" ~severity:Error;
  Reporter.emit ~loc:(Range.make (~@ s2 1 3, ~@ s2 3 4)) Hello "this is a warning" ~severity:Warning;
  Reporter.emit ~loc:(Range.make (~@ s2 1 3, ~@ s2 3 4)) Hello "this is an info" ~severity:Info;
  Reporter.emit ~loc:(Range.make (~@ s2 1 3, ~@ s2 3 4)) Hello "this is a hint" ~severity:Hint;

  Reporter.trace "When stepping into the abyss..." begin fun () ->
    Reporter.emit ~loc:(Range.make (~@ s2 1 3, ~@ s2 3 4)) Hello "Hello again!";
    Reporter.trace "When stepping into the deep abyss..." @@ fun () ->
    begin
      Reporter.emit ~loc:(Range.make (~@ s2 1 3, ~@ s2 3 4)) Hello "Hello once again!";
    end
  end;

  Reporter.trace ~loc:(Range.make (~@ s2 1 1, ~@ s2 3 4)) "When stepping into the abyss..." begin fun () ->
    Reporter.emit ~loc:(Range.make (~@ s2 1 3, ~@ s2 3 4)) Hello "Hello again!";
    Reporter.trace "When stepping into the deep abyss..." begin fun () ->
      Reporter.emit ~loc:(Range.make (~@ s2 1 3, ~@ s2 3 4)) Hello "Hello once again!" ~severity:Info;
    end
  end;

  Reporter.emit ~loc:(Range.make (~@ s1 2 3, ~@ s1 2 7)) Hello "hello here!"
    ~extra_remarks:[
      Diagnostic.loctext "Message 1";
      Diagnostic.loctext ~loc:(Range.make (~@ s2 1 3, ~@ s2 3 4)) "Message 2";
      Diagnostic.loctext "Message 3";
      Diagnostic.loctext ~loc:(Range.make (~@ s1 3 4, ~@ s1 3 5)) "Message 4";
      Diagnostic.loctext ~loc:(Range.make (~@ s1 2 8, ~@ s1 2 9)) "Message 5";
      Diagnostic.loctext ~loc:(Range.make (~@ s1 1 3, ~@ s1 2 1)) "Message 6";
      Diagnostic.loctext ~loc:(Range.make (~@ s2 1 3, ~@ s2 2 1)) "Message 7";
      Diagnostic.loctext ~loc:(Range.make (~@ s2 10 0, ~@ s2 10 0)) "Message 8";
    ];
