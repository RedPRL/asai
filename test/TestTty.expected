[0m [0mï¿«[0m [0;33mwarning[0;33m[[0;33mhello[0m[K[0;33m]E[0m [0mï¿®[0m [0m[K[0;33mthis is a warningE[0m   [0m[K[0;33mhaving two linesE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;31merror[0;31m[[0;31mbye[0m[K[0;31m]E[0m [0mï¿®[0m [0m[K[0;31mthis is an errorE[0m   [0m[K[0;31mhaving two linesE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;30;41mbug[0;30;41m[[0;30;41mhello[0m[K[0;30;41m]E[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;30;41;4maaaaaaaE[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;30;41;4mbbbbbbbbbbE[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;30;41;4mcccc[0m[K[0mccccccE[0m   [0m^[0m [0m[K[0;30;41mthis is a bugE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;31merror[0;31m[[0;31mhello[0m[K[0;31m]E[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;31;4maaaaaaaE[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;31;4mbbbbbbbbbbE[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;31;4mcccc[0m[K[0mccccccE[0m   [0m^[0m [0m[K[0;31mthis is an errorE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;33mwarning[0;33m[[0;33mhello[0m[K[0;33m]E[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;33;4maaaaaaaE[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;33;4mbbbbbbbbbbE[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;33;4mcccc[0m[K[0mccccccE[0m   [0m^[0m [0m[K[0;33mthis is a warningE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;32minfo[0;32m[[0;32mhello[0m[K[0;32m]E[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;32;4maaaaaaaE[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;32;4mbbbbbbbbbbE[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;32;4mcccc[0m[K[0mccccccE[0m   [0m^[0m [0m[K[0;32mthis is an infoE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;34mhint[0;34m[[0;34mhello[0m[K[0;34m]E[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;34;4maaaaaaaE[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;34;4mbbbbbbbbbbE[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;34;4mcccc[0m[K[0mccccccE[0m   [0m^[0m [0m[K[0;34mthis is a hintE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;33mwarning[0;33m[[0;33mhello[0m[K[0;33m]E[0m [0;38;5;240mâ•¯[0m [0mï¿®[0m [0m[K[0;33mwhen peaking into the abyssE[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;33;4maaaaaaaE[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;33;4mbbbbbbbbbbE[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;33;4mcccc[0m[K[0mccccccE[0m   [0m^[0m [0m[K[0;33mhello from hereE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;33mwarning[0;33m[[0;33mhello[0m[K[0;33m]E[0m [0;38;5;240mâ”†[0m [0mï¿®[0m [0m[K[0;33mwhen peaking into the abyssE[0m [0;38;5;240mâ•¯[0m [0mï¿®[0m [0m[K[0;33mwhen peaking into the deep abyssE[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;33;4maaaaaaaE[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;33;4mbbbbbbbbbbE[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;33;4mcccc[0m[K[0mccccccE[0m   [0m^[0m [0m[K[0;33mhello from hereE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;33mwarning[0;33m[[0;33mhello[0m[K[0;33m]E[0m [0;38;5;240mâ”†[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240mâ”†[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0ma[0m[K[0;33;4maaaaaaaaaE[0m [0;38;5;240mâ”†[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;33;4mbbbbbbbbbbE[0m [0;38;5;240mâ”†[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;33;4mcccc[0m[K[0mccccccE[0m [0;38;5;240mâ•¯[0m   [0m^[0m [0m[K[0;33mwhen stepping into the abyssE[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;33;4maaaaaaaE[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;33;4mbbbbbbbbbbE[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;33;4mcccc[0m[K[0mccccccE[0m   [0m^[0m [0m[K[0;33mhello from hereE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;32minfo[0;32m[[0;32mhello[0m[K[0;32m]E[0m [0;38;5;240mâ”†[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240mâ”†[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0ma[0m[K[0;32;4maaaaaaaaaE[0m [0;38;5;240mâ”†[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;32;4mbbbbbbbbbbE[0m [0;38;5;240mâ”†[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;32;4mcccc[0m[K[0mccccccE[0m [0;38;5;240mâ”†[0m   [0m^[0m [0m[K[0;32mwhen stepping into the abyssE[0m [0;38;5;240mâ”†[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240mâ”†[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0ma[0m[K[0;32;4maaaaaaaaaE[0m [0;38;5;240mâ”†[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;32;4mbbbbbbbbbbE[0m [0;38;5;240mâ”†[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;32;4mcccc[0m[K[0mccccccE[0m [0;38;5;240mâ•¯[0m   [0m^[0m [0m[K[0;32mwhen stepping into the deep abyssE[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;32;4maaaaaaaE[0m [0;38;5;240m2[0m [0;38;5;240m|[0m [0m[K[0;32;4mbbbbbbbbbbE[0m [0;38;5;240m3[0m [0;38;5;240m|[0m [0;32;4mcccc[0m[K[0mccccccE[0m   [0m^[0m [0m[K[0;32mhello from hereE[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;33mwarning[0;33m[[0;33mhello[0m[K[0;33m]E[0m  [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;4maaaaaaaE[0m  [0;38;5;240m2[0m [0;38;5;240m|[0m [0;4mb[0mbb[0;33;4mbbbb[0mb[0;4mb[0m[K[0mbE[0m    [0m^[0m [0m[K[0mmessage 6E[0m    [0m^[0m [0m[K[0;33mthis is the main messageE[0m    [0m^[0m [0m[K[0mmessage 5E[0m  [0;38;5;240m3[0m [0;38;5;240m|[0m [0mcccc[0;4mc[0m[K[0mcccccE[0m    [0m^[0m [0m[K[0mmessage 4E[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m  [0;38;5;240m1[0m [0;38;5;240m|[0m [0maaa[0m[K[0;4maaaaaaaE[0m  [0;38;5;240m2[0m [0;38;5;240m|[0m [0;4mb[0m[K[0;4mbbbbbbbbbE[0m    [0m^[0m [0m[K[0mmessage 7E[0m  [0;38;5;240m3[0m [0;38;5;240m|[0m [0;4mcccc[0m[K[0mccccccE[0m    [0m^[0m [0m[K[0mmessage 2E[0m [0;38;5;240m10[0m [0;38;5;240m|[0m [0m[K[0mjjjjjjjjjjE[0m    [0m^[0m [0m[K[0mmessage 8E[0m [0mï¿®[0m [0m[K[0mmessage 1E[0m [0mï¿®[0m [0m[K[0mmessage 3E[0m[KE[0m[K[0m[0m [0mï¿«[0m [0;33mwarning[0;33m[[0;33mhello[0m[K[0;33m]E[0m [0;38;5;240m23[0m [0;38;5;240m|[0m [0mwwwwwwwwww[0m[K[0;33;4mâ€¹EOFâ€ºE[0m    [0m^[0m [0m[K[0;33mthis is the main messageE[0m [0mï¿­[0m [0m[K[0m/path/to/file.coolE[0m [0;38;5;240m23[0m [0;38;5;240m|[0m [0mwwwwwwwwww[0m[K[0;4mâ€¹EOFâ€ºE[0m    [0m^[0m [0m[K[0mending of another fileE[0m[KE[0m[K[0m ï¿« warning[hello]
 ï¿® this is a warning
   having two lines

 ï¿« error[bye]
 ï¿® this is an error
   having two lines

 ï¿« bug[hello]
 ï¿­ /path/to/file.cool
 1 | aaaaaaaaaa
 2 | bbbbbbbbbb
 3 | cccccccccc
   ^ this is a bug

 ï¿« error[hello]
 ï¿­ /path/to/file.cool
 1 | aaaaaaaaaa
 2 | bbbbbbbbbb
 3 | cccccccccc
   ^ this is an error

 ï¿« warning[hello]
 ï¿­ /path/to/file.cool
 1 | aaaaaaaaaa
 2 | bbbbbbbbbb
 3 | cccccccccc
   ^ this is a warning

 ï¿« info[hello]
 ï¿­ /path/to/file.cool
 1 | aaaaaaaaaa
 2 | bbbbbbbbbb
 3 | cccccccccc
   ^ this is an info

 ï¿« hint[hello]
 ï¿­ /path/to/file.cool
 1 | aaaaaaaaaa
 2 | bbbbbbbbbb
 3 | cccccccccc
   ^ this is a hint

 ï¿« warning[hello]
 â•¯ ï¿® when peaking into the abyss
 ï¿­ /path/to/file.cool
 1 | aaaaaaaaaa
 2 | bbbbbbbbbb
 3 | cccccccccc
   ^ hello from here

 ï¿« warning[hello]
 â”† ï¿® when peaking into the abyss
 â•¯ ï¿® when peaking into the deep abyss
 ï¿­ /path/to/file.cool
 1 | aaaaaaaaaa
 2 | bbbbbbbbbb
 3 | cccccccccc
   ^ hello from here

 ï¿« warning[hello]
 â”† ï¿­ /path/to/file.cool
 â”† 1 | aaaaaaaaaa
 â”† 2 | bbbbbbbbbb
 â”† 3 | cccccccccc
 â•¯   ^ when stepping into the abyss
 ï¿­ /path/to/file.cool
 1 | aaaaaaaaaa
 2 | bbbbbbbbbb
 3 | cccccccccc
   ^ hello from here

 ï¿« info[hello]
 â”† ï¿­ /path/to/file.cool
 â”† 1 | aaaaaaaaaa
 â”† 2 | bbbbbbbbbb
 â”† 3 | cccccccccc
 â”†   ^ when stepping into the abyss
 â”† ï¿­ /path/to/file.cool
 â”† 1 | aaaaaaaaaa
 â”† 2 | bbbbbbbbbb
 â”† 3 | cccccccccc
 â•¯   ^ when stepping into the deep abyss
 ï¿­ /path/to/file.cool
 1 | aaaaaaaaaa
 2 | bbbbbbbbbb
 3 | cccccccccc
   ^ hello from here

 ï¿« warning[hello]
  1 | aaaaaaaaaa
  2 | bbbbbbbbbb
    ^ message 6
    ^ this is the main message
    ^ message 5
  3 | cccccccccc
    ^ message 4
 ï¿­ /path/to/file.cool
  1 | aaaaaaaaaa
  2 | bbbbbbbbbb
    ^ message 7
  3 | cccccccccc
    ^ message 2
 10 | jjjjjjjjjj
    ^ message 8
 ï¿® message 1
 ï¿® message 3

 ï¿« warning[hello]
 23 | wwwwwwwwwwâ€¹EOFâ€º
    ^ this is the main message
 ï¿­ /path/to/file.cool
 23 | wwwwwwwwwwâ€¹EOFâ€º
    ^ ending of another file

